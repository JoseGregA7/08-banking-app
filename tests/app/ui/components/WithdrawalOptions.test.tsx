import { render, fireEvent } from '@testing-library/react';
import React from 'react';
import '@testing-library/jest-dom';
import { describe, expect, test, vitest } from 'vitest';
import WithdrawOptions from '../../../../src/app/ui/components/WithdrawalOptions';

describe('WithdrawOptions', () => {
    const mockHandleWithdrawTypeChange = vitest.fn();
    const mockHandleWithdrawalAmountChange = vitest.fn();
    const mockHandleWithdrawalSubmit = vitest.fn();
    const state = {
        withdrawalType: 'ATM_WITHDRAWAL',
        withdrawalAmount: 100,
    };

    test('should call handleWithdrawalSubmit when submit button is clicked', () => {
        const { getByTestId } = render(
            <WithdrawOptions
                handleWithdrawTypeChange={mockHandleWithdrawTypeChange}
                handleWithdrawalAmountChange={mockHandleWithdrawalAmountChange}
                handleWithdrawalSubmit={mockHandleWithdrawalSubmit}
                state={state}
            />
        );

        const formElement = getByTestId('withdraw-options-wrapper');
        expect(formElement).toBeInTheDocument();

        fireEvent.click(getByTestId('withdraw-submit-button'));

        expect(mockHandleWithdrawalSubmit).toHaveBeenCalledTimes(1);
    });

    test('should call handleWithdrawTypeChange when withdrawal type is selected', () => {
        const { getByTestId } = render(
            <WithdrawOptions
                handleWithdrawTypeChange={mockHandleWithdrawTypeChange}
                handleWithdrawalAmountChange={mockHandleWithdrawalAmountChange}
                handleWithdrawalSubmit={mockHandleWithdrawalSubmit}
                state={state}
            />
        );

        fireEvent.change(getByTestId('withdraw-type-select'), {
            target: { value: 'OTHER_WITHDRAWAL' },
        });

        expect(mockHandleWithdrawTypeChange).toHaveBeenCalledTimes(1);
    });

    test('should call handleWithdrawalAmountChange when amount is entered', () => {
        const { getByTestId } = render(
            <WithdrawOptions
                handleWithdrawTypeChange={mockHandleWithdrawTypeChange}
                handleWithdrawalAmountChange={mockHandleWithdrawalAmountChange}
                handleWithdrawalSubmit={mockHandleWithdrawalSubmit}
                state={state}
            />
        );

        fireEvent.change(getByTestId('withdraw-amount-input'), {
            target: { value: '200' },
        });

        expect(mockHandleWithdrawalAmountChange).toHaveBeenCalledTimes(1);
    });

    test('should match snapshot', () => {
        const { asFragment } = render(
            <WithdrawOptions
                handleWithdrawTypeChange={mockHandleWithdrawTypeChange}
                handleWithdrawalAmountChange={mockHandleWithdrawalAmountChange}
                handleWithdrawalSubmit={mockHandleWithdrawalSubmit}
                state={state}
            />
        );

        expect(asFragment()).toMatchSnapshot();
    });
});
