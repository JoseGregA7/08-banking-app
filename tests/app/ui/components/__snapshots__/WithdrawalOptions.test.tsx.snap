// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WithdrawOptions > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="WithdrawOptions__wrapper"
    data-testid="withdraw-options-wrapper"
  >
    <h3
      class="WithdrawOptions__title"
      id="withdraw-form-title"
    >
      Formulario de Retiro
    </h3>
    <form
      aria-labelledby="withdraw-form-title"
      class="WithdrawOptions__form"
    >
      <div
        class="WithdrawOptions__field"
      >
        <label
          class="WithdrawOptions__label"
          for="withdrawType"
        >
          Tipo de Retiro:
        </label>
        <select
          aria-label="Seleccionar el tipo de retiro"
          class="WithdrawOptions__select"
          data-testid="withdraw-type-select"
          id="withdrawType"
        >
          <option
            value="ATM_WITHDRAWAL"
          >
            Retiro desde Cajero
          </option>
          <option
            value="OTHER_WITHDRAWAL"
          >
            Otro tipo de retiro
          </option>
        </select>
      </div>
      <div
        class="WithdrawOptions__field"
      >
        <label
          class="WithdrawOptions__label"
          for="withdrawalAmount"
        >
          Cantidad a retirar:
        </label>
        <input
          aria-label="Cantidad a retirar"
          class="WithdrawOptions__input"
          data-testid="withdraw-amount-input"
          id="withdrawalAmount"
          min="0"
          type="number"
          value="100"
        />
      </div>
      <button
        aria-label="Realizar retiro"
        class="WithdrawOptions__submitButton"
        data-testid="withdraw-submit-button"
        type="submit"
      >
        Realizar Retiro
      </button>
    </form>
  </div>
</DocumentFragment>
`;
